common/geth/multihoming_client.go

```diff
@@ -0,0 +1,707 @@
+package geth
+
+import (
+	"context"
+	"math/big"
+
+	dacommon "github.com/Layr-Labs/eigenda/common"
+	"github.com/Layr-Labs/eigensdk-go/logging"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	gethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+type MultiHomingClient struct {
+	RPCs       []dacommon.EthClient
+	rpcUrls    []string
+	NumRetries int
+	Logger     logging.Logger
+	*FailoverController
+}
+
+var _ dacommon.EthClient = (*MultiHomingClient)(nil)
+
+// NewMultiHomingClient is an EthClient that automatically handles RPC failures and retries by cycling through
+// multiple RPC clients. All EthClients underneath maintain active connections throughout the life time. The
+// MultiHomingClient keeps using the same EthClient for a new RPC invocation until it encounters a connection
+// error (i.e. any Non EVM error). Then the next EthClient is chosen in a round robin fashion, and the same rpc call
+// can be retried. The total number of retry is configured through cli argument. When the rpc call has used up all
+// the retry opportunity, the rpc would fail and return error. The MultiHomingClient assumes a single private key.
+func NewMultiHomingClient(config EthClientConfig, logger logging.Logger) (*MultiHomingClient, error) {
+	rpcUrls := config.RPCURLs
+
+	controller := NewFailoverController(len(rpcUrls), logger)
+
+	client := &MultiHomingClient{
+		rpcUrls:            rpcUrls,
+		NumRetries:         config.NumRetries,
+		FailoverController: controller,
+		Logger:             logger,
+	}
+
+	for i := 0; i < len(rpcUrls); i++ {
+		rpc, err := NewClient(config, i, logger)
+		if err != nil {
+			logger.Info("cannot connect to rpc at start", "url", rpcUrls[i])
+			return nil, err
+		}
+		client.RPCs = append(client.RPCs, rpc)
+	}
+
+	return client, nil
+}
+
+func (m *MultiHomingClient) GetRPCInstance() (int, dacommon.EthClient) {
+	index := m.GetTotalNumberRpcFault() % uint64(len(m.RPCs))
+	return int(index), m.RPCs[index]
+}
+
+func (m *MultiHomingClient) GetAccountAddress() gethcommon.Address {
+	_, instance := m.GetRPCInstance()
+	return instance.GetAccountAddress()
+}
+
+func (m *MultiHomingClient) SuggestGasTipCap(ctx context.Context) (*big.Int, error) {
+	var errLast error
+	for i := 0; i < m.NumRetries+1; i++ {
```

Comment:
This block of code is repeating in each function, can it be pulled out and shared? It looks just need to wrap the `instance.SomeFunc(SomeArgs)` part.

Reply:
I don't think ethclient from geth support grpc, only jsonRpc. https://geth.ethereum.org/docs/interacting-with-geth/rpc